File: Python/balancedBinaryTree.py
# O(N^2)
class Solution:
    def height(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        left = self.height(root.left)
        right = self.height(root.right)
        return max(left, right) + 1

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
        l = self.height(root.left)
        r = self.height(root.right)
        if abs(l - r) > 1:
            return False
        left = self.isBalanced(root.left)
        right = self.isBalanced(root.right)
        return left and right

# O(N)

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self.height(root) != -1

    def height(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        l = self.height(root.left)
        if l == -1:
            return -1
        r = self.height(root.right)
        if r == -1:
            return -1
        if abs(l - r) > 1:
            return -1
        return max(l, r) + 1

File: Java/balancedBinaryTree.java
// O(N^2) solution
class Solution1 {

    public int height(TreeNode root) {
        if (root == null)
            return 0;

        int left = height(root.left);
        int right = height(root.right);

        return Math.max(left, right) + 1;
    }

    public boolean isBalanced(TreeNode root) {
        if (root == null)
            return true;

        int l = height(root.left); 
        int r = height(root.right); 

        if (Math.abs(l - r) > 1)
            return false;

        boolean left = isBalanced(root.left);
        boolean right = isBalanced(root.right);

        return left && right;
    }
}

//O(N) solution
class Solution {
    public boolean isBalanced(TreeNode root) {
        return height(root)!=-1;   
    }

    public int height(TreeNode root){
        if(root==null)
            return 0;
        
        int l= height(root.left);
        if(l==-1)
            return -1;

        int r= height(root.right);
        if(r==-1)
            return -1;

        if(Math.abs(l-r)>1)
            return -1;
        
        return Math.max(l,r)+1;
    }
}